src/networkapi.jl

```diff
@@ -465,6 +465,68 @@ function complexoutgoingmat(rn::ReactionSystem; sparse=false, B=reactioncomplexe
 end
 
 
+"""
+   incidencematgraph(incidencemat::Matrix)
+   incidencematgraph(incidencemat::SparseMatrixCSC{Int,Int})
+
+Given an incidence matrix of reaction-network in dense or sparse matrix form,
+writtens a {n ,m} directed simple graph,
+where n is number of nodes(reaction-complexes)
+and m is number of edges(reactions)
+"""
+function incidencematgraph(incidencemat::Matrix{Int})
+   @assert all(∈([-1,0,1]) ,incidencemat)
+   n = size(incidencemat,1)  # no. of nodes/complexes
+   graph = LightGraphs.DiGraph(n)
+   for col in eachcol(incidencemat)
+      add_edge!(graph, findmin(col)[2],findmax(col)[2])
+   end
+   return graph
+end
+function incidencematgraph(incidencemat::SparseMatrixCSC{Int,Int})
+   @assert all(∈([-1,0,1]) ,incidencemat)
+   m,n = size(incidencemat)  
+   graph = LightGraphs.DiGraph(m)
+   rows = rowvals(incidencemat)
+   vals = nonzeros(incidencemat)
+   for j = 1:n
+      inds=nzrange(incidencemat, j)
+      row = rows[inds];
+      val = vals[inds];
+      if val[1] == -1
+         add_edge!(graph, row[1], row[2])
+      else
+         add_edge!(graph, row[2], row[1])
+      end
+   end
+   return graph
+end
+
+
+"""
+Given a directed simple graph ,created from incidence matrix
+of reaction network, returns Vector of indices of directly connected
+reaction-complexes(more precisely indices of reactioncomplexes(network)[1])
+in reactions
+"""
+function linkageclasses(graph::SimpleDiGraph{Int64})
+   LightGraphs.connected_components(graph)
+end
+
+
+"""
+Given a ReactionSystem, returns deficiency of the reaction network
+defined as,
+deficiency = number of reaction complexes - number of linkage classes - dimension of stochiometric subspace
+"""
+function deficiency(rs::ReactionSystem;sparse::Bool=false)
```

Comment:
```suggestion
function deficiency(rs::ReactionSystem, ns, ig, lc=linkageclasses(ig))
```
where `ns` is the netstoichiometry, `ig` is the incidence graph, and `lc` are the linkage classes. You can get the number of complexes from the number of nodes in `ig` right? (There should be a Lightgraphs function for that.)
```

Reply:
@isaacsas  yes, number of vetrtices cn be found using `nv` function in LightGraphs
how about following , as per other functions we have written previously in API ? 

```julia
function deficiency(rs::ReactionSystem;sparse::Bool=false,ns=netstoichmat(rs;sparse),
                        ig = incidencematgraph(reactioncomplexes(rs;sparse=sparse)[2]),
                        lc=linkageclasses(ig))

   LightGraphs.nv(ig) - length(lc) - AbstractAlgebra.rank(AbstractAlgebra.matrix(AbstractAlgebra.ZZ,ns))
end

```

Reply:
> @isaacsas yes, number of vertices can be found using `nv` function in LightGraphs
> how about following , as per other functions we have written previously in API ?
> We need `reactionSystem` to be only input,. rest all as optional kwargs supplied as below or user can pass them in case they have been pre-calculated
> 
> ```julia
> function deficiency(rs::ReactionSystem ; sparse::Bool=false, ns::AbstractMatrix = netstoichmat(rs;sparse),
>                         ig::SimpleDiGraph{Int64} = incidencematgraph(reactioncomplexes(rs;sparse=sparse)[2]),
>                         lc::Vector{Vector{Int64}} = linkageclasses(ig))
> 
>    LightGraphs.nv(ig) - length(lc) - AbstractAlgebra.rank(AbstractAlgebra.matrix(AbstractAlgebra.ZZ,ns))
> end
> ```

@isaacsas , i myself am now in doubt of this..  if `ns`, `ig` and `lc` are precalculated and supplied by user , `rs::ReactionSystem` remains unused !